README.md
## pagedata-api


CHANGELOG.md

0.5.0 / 2016-12-02
==================

  * default user-agent is pagedata-api/${version}
  * removed outdated update method
  * added userAgent

0.4.0 / 2016-10-20
==================

  * updates for new version of pagedata

0.3.0 / 2016-06-22
==================

  * better error responses

0.2.0 / 2016-06-13
==================

  * changed tag to be passed on individual calls

0.1.1 / 2016-06-13
==================

  * default tag to blank

0.1.0 / 2016-06-13
==================

  * updated for new version with tags

0.0.3 / 2016-06-13
==================

  * added getManyAndMerge to get multiple pages and merge them together

0.0.2 / 2016-05-26
==================

  * added support if content came back as a string


test/main.test.js
/* eslint max-len:0, no-console: 0, no-unused-vars: 0, strict: 0 */
'use strict';

const Code = require('code');
const Lab = require('lab');
const lab = exports.lab = Lab.script();

const sinon = require('sinon');

const SiteInfoAPI = require('../');
const server = require('./helpers/test_api_server.js');


lab.experiment('PageDataAPI', () => {
  lab.experiment('#()', () => {
    lab.test('it should return and object upon instatiation', done => {
      const siteInfo = new SiteInfoAPI('http://apiurl.com', 'dev', 'somerandomkey');
      Code.expect(siteInfo).to.be.an.object();

      done();
    });
  });

  lab.experiment('.get', () => {
    lab.before(done => {
      server.listen(3080, done);
    });

    lab.after(done => {
      server.close();
      done();
    });

    lab.test('it should call back and error if the slug is not passed', done => {
      const siteInfo = new SiteInfoAPI('http://apiurl', 'dev', 'somekey');
      const errCb = sinon.spy();

      siteInfo.get(null, errCb);

      Code.expect(errCb.args[0][0]).to.exist();
      Code.expect(errCb.args[0][0]).to.be.a.string();
      done();
    });

    lab.test('it should attempt to get an page from the api', done => {
      const siteInfo = new SiteInfoAPI('http://localhost:3080', 'dev', 'na');
      siteInfo.get('some-page', (err, result) => {
        Code.expect(err).to.be.null();
        Code.expect(result).to.be.an.object();
        done();
      });
    });
  });
});


tests/test.api.js
'use strict';
const code = require('code');   // assertion library
const Lab = require('lab');
const lab = exports.lab = Lab.script();
const Hapi = require('hapi');
const PageData = require('../index.js');

lab.experiment('hapi-auth-email hashing', () => {
  let server;
  const url = 'http://localhost:8000';
  const key = 'theKey';
  const userAgent = '007';
  lab.beforeEach((done) => {
    server = new Hapi.Server();
    server.connection({ host: 'localhost', port: 8000 });
    server.start(done);
  });

  lab.afterEach((done) => {
    server.stop(done);
  });

  lab.test('can instantiate the PageData API', (done) => {
    const pageData = new PageData(url, key, userAgent);
    code.expect(typeof pageData).to.equal('object');
    code.expect(typeof pageData.getPages).to.equal('function');
    code.expect(typeof pageData.getUrl).to.equal('function');
    code.expect(typeof pageData.get).to.equal('function');
    code.expect(typeof pageData.getMany).to.equal('function');
    code.expect(typeof pageData.getManyAndMerge).to.equal('function');
    code.expect(pageData.options.url).to.equal(url);
    code.expect(pageData.options.key).to.equal(key);
    code.expect(pageData.options.userAgent).to.equal(userAgent);
    done();
  });

  lab.test('getPages', (done) => {
    server.route({
      path: '/api/sites/{site}/pages',
      method: 'GET',
      handler: (request, reply) => {
        code.expect(request.headers).to.include('user-agent');
        code.expect(request.headers['user-agent']).to.equal(userAgent);
        code.expect(request.headers).to.include('x-api-key');
        code.expect(request.headers['x-api-key']).to.equal(key);
        reply({ hello: 'world' });
      }
    });
    const pageData = new PageData(url, key, userAgent);
    pageData.getPages('mySite', (err, result) => {
      if (err) {
        throw err;
      }
      code.expect(result.hello).to.equal('world');
      done();
    });
  });

  lab.test('get', (done) => {
    server.route({
      path: '/api/sites/{site}/pages/{slug}',
      method: 'GET',
      handler: (request, reply) => {
        code.expect(request.headers).to.include('user-agent');
        code.expect(request.headers).to.include('x-api-key');
        code.expect(request.params.site).to.equal('mySite');
        code.expect(request.params.slug).to.equal('mySlug');
        reply({ payload: { hello: 'world' } });
      }
    });
    const pageData = new PageData(url, key, userAgent);
    pageData.get('mySite', 'mySlug', 'myTag', (err, result) => {
      if (err) {
        throw err;
      }
      code.expect(result.payload.hello).to.equal('world');
      done();
    });
  });

  lab.test('getMany', { timeout: 5000 }, (done) => {
    const tag = 'myTag';
    const slugs = ['my-slug-1', 'my-slug', 'another-slug'];

    server.route({
      path: '/api/sites/{site}/pages/{slug}',
      method: 'GET',
      handler: (request, reply) => {
        code.expect(request.headers).to.include('user-agent');
        code.expect(request.headers).to.include('x-api-key');
        code.expect(tag).to.equal(request.query.tag);
        code.expect(slugs).to.include(request.params.slug);
        reply({ payload: { hello: 'world' } });
      }
    });
    const pageData = new PageData(url, key, userAgent);
    pageData.getMany('mySite', slugs, tag, (err, result) => {
      if (err) {
        throw err;
      }
      code.expect(result.length).to.equal(slugs.length);
      code.expect(result[0].payload.hello).to.equal('world');
      done();
    });
  });

  lab.test('getManyAndMerge', { timeout: 5000 }, (done) => {
    const tag = 'myTag';
    const slugs = ['my-slug-1', 'my-slug', 'another-slug'];

    server.route({
      path: '/api/sites/{site}/pages/{slug}',
      method: 'GET',
      handler: (request, reply) => {
        code.expect(request.headers).to.include('user-agent');
        code.expect(request.headers).to.include('x-api-key');
        code.expect(tag).to.equal(request.query.tag);
        code.expect(slugs).to.include(request.params.slug);
        reply({ payload: { hello: 'world' } });
      }
    });
    const pageData = new PageData(url, key, userAgent);
    pageData.getMany('mySite', slugs, tag, (err, result) => {
      if (err) {
        throw err;
      }
      code.expect(result.length).to.equal(slugs.length);
      code.expect(result[0].payload.hello).to.equal('world');
      done();
    });
  });

  lab.test('userAgent is optional', (done) => {
    const defaultUserAgent = `pagedata-api/${require('../package.json').version}`;
    server.route({
      path: '/api/sites/{site}/pages',
      method: 'GET',
      handler: (request, reply) => {
        code.expect(request.headers).to.include('user-agent');
        code.expect(request.headers['user-agent']).to.equal(defaultUserAgent);
        code.expect(request.headers).to.include('x-api-key');
        code.expect(request.headers['x-api-key']).to.equal(key);
        reply({ hello: 'world' });
      }
    });
    const pageData = new PageData(url, key);
    pageData.getPages('mySite', (err, result) => {
      if (err) {
        throw err;
      }
      code.expect(result.hello).to.equal('world');
      done();
    });
  });
});


test/helpers/test_api_server.js


const http = require('http');

module.exports = http.createServer((req, res) => {
  if (req.url === '/dev/api/page/some-page') {
    res.writeHead(200, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify({ content: { one: 'uno', two: 'dos' } }));
  } else {
    res.writeHead(404, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify({ err: 'No such luck', url: req.url }));
  }
});


lib/base.js
'use strict';

const wreck = require('wreck');

class PageData {
  constructor(url, env, key) {
   this.options = {
    url: url,
    env: env,
    key: key
   }; 
  
    return this;
  }

  getUrl(slug) {
    return `${this.options.url}/${this.options.env}/api/page/${slug}`;
  }

  get(slug, cb) {
    const url = this.getUrl(slug);
    wreck.get(url, {
      json: true,
      headers: {
        'x-api-key': this.options.key
      }
    }, (err, res, payload) => {
      if(err) {
        return cb(err);
      }
      if(res.statusCode !== 200) {
        return cb({message: 'Api returned a non 200 status code', statusCode: res.statusCode, result: res});
      }
      return cb(null, payload);
    });
  }

  getMany(slugs, cb) {
    const self = this;
    async.map(slugs, (slug, done) => {
      self.get(slug, done);
    }, cb); 
  }

  update(slug, content, cb) {
    const url = this.getUrl(slug);
    const headers = {
      'x-api-key': this.options.key
    };

    if(!content.content) {
      return cb({message: 'Payload must include content attirbute'});
    }

    const payload = JSON.stringify(content);

    wreck.put(url, {json: true, headers: headers, payload: payload}, (err, res, payload) => {
      if(err) {
        return cb(err);
      }
      if(res.statusCode !== 200) {
        return cb({message: 'Api returned a non 200 status code', statusCode: res.statusCode, result: res});
      }
      
      cb(null, payload);
    });
  }

}

module.exports = PageData;


index.js
'use strict';

const wreck = require('wreck');
const async = require('async');
const defaults = require('lodash.defaultsdeep');
const version = require('./package.json').version;

class PageData {
  constructor(url, key, userAgent) {
    this.options = {
      url,
      key,
      userAgent: userAgent ? userAgent : `pagedata-api/${version}`
    };
  }
  getPages(site, done) {
    const url = `${this.options.url}/api/sites/${site}/pages`;
    const headers = {
      'x-api-key': this.options.key
    };
    if (this.options.userAgent) {
      headers['user-agent'] = this.options.userAgent;
    }
    wreck.get(url, {
      json: true,
      headers,
    }, (err, res, payload) => {
      if (err) {
        return done(err);
      }
      done(null, payload);
    });
  }

  getUrl(site, slug, tag) {
    return `${this.options.url}/api/sites/${site}/pages/${slug}?tag=${tag}`;
  }

  get(site, slug, tag, cb) {
    if (typeof tag === 'function') {
      cb = tag;
      tag = '';
    }
    const url = this.getUrl(site, slug, tag);
    const headers = {
      'x-api-key': this.options.key
    };
    if (this.options.userAgent) {
      headers['user-agent'] = this.options.userAgent;
    }
    wreck.get(url, {
      json: true,
      headers,
    }, (err, res, payload) => {
      if (err) {
        return cb(err);
      }
      if (res.statusCode === 404) {
        return cb(new Error('Invalid slug and/or tag'));
      }
      if (res.statusCode !== 200) {
        return cb({ message: 'Api returned a non 200 status code', statusCode: res.statusCode, payload });
      }
      if (typeof payload.content === 'string') {
        try {
          payload.content = JSON.parse(payload.content);
        } catch (e) {
          return cb(e);
        }
      }
      return cb(null, payload);
    });
  }

  getMany(site, slugs, tag, cb) {
    async.map(slugs, (slug, done) => {
      this.get(site, slug, tag, done);
    }, cb);
  }

  getManyAndMerge(site, slugs, tag, cb) {
    async.map(slugs, (slug, done) => {
      this.get(site, slug, tag, done);
    }, (err, data) => {
      if (err) {
        return cb(err);
      }
      const content = data.map(item => item.content);
      content.unshift({});
      const merged = defaults.apply(null, content);
      cb(null, merged);
    });
  }

}

module.exports = PageData;
